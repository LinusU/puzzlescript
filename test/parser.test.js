const {parse, parseGrammar} = require('../src/parser')

const OBJECT_HEADER = `
===
OBJECTS
===
`

const RULE_HEADER = `
===
RULES
===
`

function checkGrammar(code) {
  // add a header for the parser
  code = `title checkGrammar\n${RULE_HEADER}\n${code}`

  const {match, grammar} = parseGrammar(code)
  if (!match.succeeded()) {
    const trace = grammar.trace(code)
    console.log(trace.toString())
  }
  expect(match.succeeded()).toBe(true)

  const s = grammar.createSemantics()
  s.addOperation('toJSON2', {
    _terminal: function () { return this.primitiveValue },
    _iter: function(children) {
      return children.map(child => child.toJSON2())
    },
    _default: function(children) {
      if (this.ctorName === 'word') {
        return this.sourceString
      // } if (this.ctorName[0] === this.ctorName[0].toLowerCase()) {
      //   return this.ctorName
      } else {
        const obj = {
          __name: this.ctorName
        }
        children.forEach((child, index) => {
          const value = child.toJSON2()
          if (!Array.isArray(value) || value.length >= 1) {
            obj[`_i${index}`] = value
          }
        })
        return obj
      }
    }
  })
  const tree = s(match).toJSON2()
  expect(tree).toMatchSnapshot()
  return tree
}

function checkParse(code, varNames) {
  // Now check if the semantics parsed
  const legendItems = varNames.map(varName => {
    return `${varName} = testObject`
  })
  const {data, error, trace} = parse(`
title checkParse

===
OBJECTS
===

testObject
transparent

===
LEGEND
===

${legendItems.join('\n')}

====
RULES
====

${code}
`)
  if (error) {
    console.log(error.message)
  }
  expect(error).toBeFalsy()
  expect(data).toMatchSnapshot()
  return data
}

function parseRule(code, varNames) {
  checkGrammar(code)
  return checkParse(code, varNames)
}


describe('rules', () => {
  it('parses a simple rule', () => {
    parseRule('[ z ] -> [ ]', ['z'])
  })
  it('parses a simple rule 2', () => {
    parseRule('[ z ] -> [ z ]', ['z'])
  })
  it('parses a simple rule without whitespace', () => {
    parseRule('[z]->[z]', ['z'])
  })
  it('parses a rule with multiple cells', () => {
    parseRule('[ z | x ] -> [ ]', ['z', 'x'])
  })
  it('parses a rule with multiple layers', () => {
    parseRule('[ z x ] -> [ ]', ['z', 'x'])
  })
  it('parses a rule with ellpisis', () => {
    parseRule('[ z | ... | x ] -> [ RANDOM z | ... | x ]', ['z', 'x'])
  })
  it('parses a rule with a period for a variable name', () => {
    parseRule('[.] -> []', ['.'])
  })
  it('parses a rule with a _ for a variable name', () => {
    parseRule('[z_x] -> []', ['z_x'])
  })


  describe('Cell Modifiers', () => {
    it('parses a rule with directions', () => {
      parseRule('[ACTION z ] -> [ ]', ['z'])
      parseRule('[^ z ] -> [ ]', ['z'])
      parseRule('[v z ] -> [ ]', ['z']) // This needs to be the down arrow, not a variable
      parseRule('[> z ] -> [ ]', ['z'])
      parseRule('[< z ] -> [ ]', ['z'])
      parseRule('[LEFT z ] -> [ ]', ['z'])
      parseRule('[RIGHT z ] -> [ ]', ['z'])
      parseRule('[UP z ] -> [ ]', ['z'])
      parseRule('[DOWN z ] -> [ ]', ['z'])
      parseRule('[STATIONARY z ] -> [ ]', ['z'])
      parseRule('[MOVING z ] -> [ ]', ['z'])
      parseRule('[VERTICAL z ] -> [ ]', ['z'])
      parseRule('[HORIZONTAL z ] -> [ ]', ['z'])
      parseRule('[PERPENDICULAR z ] -> [ ]', ['z'])
      parseRule('[ORTHOGONAL z ] -> [ ]', ['z'])
      parseRule('[RANDOMDIR z ] -> [ ]', ['z'])
    })
    it('parses a rule with modifiers', () => {
      parseRule('[ NO z ] -> [ ]', ['z'])
    })
    it('parses a rule with a variable that begins with the name of a modifier', () => {
      parseRule('[ stationaryz ] -> [ stationaryz ]', ['stationaryz'])
    })
    it('parses a rule with modifiers 2', () => {
      parseRule('[ STATIONARY z ] -> [ z ]', ['z'])
    })
  })

  describe('Commands', () => {
    it('parses a rule with a command inside the brackets (these should be moved up to the Action Commands)', () => {
      parseRule('[z]->[SFX0]', ['z'])
      parseRule('[z]->[z SFX1]', ['z'])
      parseRule('[z]->[z winter]', ['z', 'winter'])
    })
  })


})
