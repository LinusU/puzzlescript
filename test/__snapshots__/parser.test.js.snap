// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules General Formatting Correctly parses weird variable names 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [
    Object {
      "name": "Bush?",
      "sourceOffset": 28,
      "type": "LEGEND_ITEM_SIMPLE",
      "value": "Player",
    },
    Object {
      "name": ".",
      "sourceOffset": 43,
      "type": "LEGEND_ITEM_SIMPLE",
      "value": "Player",
    },
  ],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 85,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 85,
                  "tile": "Bush?",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 83,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 74,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 74,
                  "tile": "Bush?",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 73,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 73,
      "type": "RULE",
    },
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 100,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 100,
                  "tile": ".",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 98,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 93,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 93,
                  "tile": ".",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 92,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 92,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "foo",
  "winConditions": Array [],
}
`;

exports[`rules parses a rule with a _ for a variable name 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "sourceOffset": 44,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 42,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 35,
                  "tile": "z_x",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a rule with a period for a variable name 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "sourceOffset": 42,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 40,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 35,
                  "tile": ".",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a rule with ellpisis 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "afterNeighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 75,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 76,
                  "tile": "x",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
            Object {
              "sourceOffset": 79,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
          ],
          "beforeNeighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 58,
              "tilesWithModifier": Array [
                Object {
                  "modifier": "RANDOM",
                  "sourceOffset": 59,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "debugFlag": null,
          "sourceOffset": 56,
          "type": "ELLIPSIS_BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "afterNeighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 45,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 46,
                  "tile": "x",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
            Object {
              "debugFlag": null,
              "sourceOffset": 49,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 50,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "beforeNeighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 36,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "debugFlag": null,
          "sourceOffset": 34,
          "type": "ELLIPSIS_BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a rule with multiple cells 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "sourceOffset": 48,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
            Object {
              "sourceOffset": 50,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 46,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 36,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
            Object {
              "debugFlag": null,
              "sourceOffset": 39,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 40,
                  "tile": "x",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a rule with multiple layers 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "sourceOffset": 46,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 44,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 36,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
                Object {
                  "modifier": null,
                  "sourceOffset": 38,
                  "tile": "x",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a simple rule 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "sourceOffset": 44,
              "tilesWithModifier": Array [],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 42,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 36,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a simple rule 2 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 44,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 45,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 42,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 36,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;

exports[`rules parses a simple rule without whitespace 1`] = `
Object {
  "collisionLayers": Array [],
  "legendItems": Array [],
  "levels": Array [],
  "metadata": Array [],
  "rules": Array [
    Object {
      "actions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 40,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 40,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 39,
          "type": "BRACKET",
        },
      ],
      "commands": Array [],
      "conditions": Array [
        Object {
          "againHack": false,
          "debugFlag": null,
          "neighbors": Array [
            Object {
              "debugFlag": null,
              "sourceOffset": 35,
              "tilesWithModifier": Array [
                Object {
                  "modifier": null,
                  "sourceOffset": 35,
                  "tile": "z",
                  "type": "TILE_WITH_MODIFIER",
                },
              ],
              "type": "NEIGHBOR",
            },
          ],
          "sourceOffset": 34,
          "type": "BRACKET",
        },
      ],
      "modifiers": Array [],
      "sourceOffset": 34,
      "type": "RULE",
    },
  ],
  "sounds": Array [],
  "sprites": Array [],
  "title": "checkGrammar",
  "winConditions": Array [],
}
`;
